src/JsonSchema/SchemaFactory.php

```diff
@@ -210,15 +241,37 @@ private function buildPropertySchema(Schema $schema, string $definitionName, str
                 $className = $valueType->getClassName();
             }
 
-            $valueSchema = $this->typeFactory->getType(new Type($builtinType, $type->isNullable(), $className, $isCollection, $keyType, $valueType), $format, $propertyMetadata->isReadableLink(), $serializerContext, $schema);
+            $valueSchema[] = $this->typeFactory->getType(new Type($builtinType, $type->isNullable(), $className, $isCollection, $keyType, $valueType), $format, $propertyMetadata->isReadableLink(), $serializerContext, $schema);
         }
 
-        if (\array_key_exists('type', $propertySchema) && \array_key_exists('$ref', $valueSchema)) {
-            $propertySchema = new \ArrayObject($propertySchema);
-        } else {
-            $propertySchema = new \ArrayObject($propertySchema + $valueSchema);
+        $valueSchema = array_unique($valueSchema, \SORT_REGULAR);
+
+        // only one builtInType detected (should be "type" or "$ref")
+        if (1 === \count($valueSchema)) {
+            $schema->getDefinitions()[$definitionName]['properties'][$normalizedPropertyName] = new \ArrayObject($propertySchema + $valueSchema[0]);
+
+            return;
         }
-        $schema->getDefinitions()[$definitionName]['properties'][$normalizedPropertyName] = $propertySchema;
+
+        // multiple builtInTypes detected: determine anyOf/oneOf/allOf if union vs intersect types
+        $reflectionType = $reflectionProperty?->getType();
+        switch (true) {
+            default:
+                // cannot detect union/intersect types
+                $composition = 'anyOf';
+                break;
+            case $reflectionType instanceof \ReflectionUnionType:
+                $composition = 'oneOf';
+                break;
+            case $reflectionType instanceof \ReflectionIntersectionType:
+                $composition = 'allOf';
+                break;
+        }
```

Comment:
as said earlier I think that this needs to be inside the Metadata, I have an even better suggestion. The idea is to create a `PropertySchemaMetadataFactory` that would fill the `$propertySchema` with the appropriate data. This way we could cache that information within property metadata and this class would be simplified a lot.

Reply:
Could be interesting to do the same for resource, like `SchemaResourceCollectionMetadataFactory`, with code from SchemaFactory::buildSchema.

WDYT?

Reply:
Can't move all code to `SchemaPropertyMetadataFactory` because `$this->typeFactory->getType` requires `$format`, `$serializerContext` and `$schema` which are only available on SchemaFactory.

I suggest to move some code to `SchemaPropertyMetadataFactory`, but keep `$valueSchema` creation on SchemaFactory::buildPropertySchema to complete the property schema.